File: 07_linked_list/07_07_odd_even_linked_list/odd_even_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def odd_even_list(head):
    if not head or not head.next:
        return head

    odd = head
    even = head.next
    even_head = head.next
    while even and even.next:
        odd.next = odd.next.next
        even.next = even.next.next
        odd = odd.next
        even = even.next

    odd.next = even_head
    return head

# Helper function to print the linked list
def print_list(label, head):
    print(label, end=" ")
    current = head
    while current:
        print(current.val, end=" ")
        current = current.next
    print()

# Test Case 1: Odd and even number of nodes
head1 = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5, ListNode(6))))))
print_list("Test Case 1 (Before):", head1)
result1 = odd_even_list(head1)
print_list("Test Case 1 (After):", result1)  # Output: 1 -> 3 -> 5 -> 2 -> 4 -> 6

# Test Case 2: Odd number of nodes
head2 = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
print_list("Test Case 2 (Before):", head2)
result2 = odd_even_list(head2)
print_list("Test Case 2 (After):", result2)  # Output: 1 -> 3 -> 5 -> 2 -> 4

# Test Case 3: Even number of nodes
head3 = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
print_list("Test Case 3 (Before):", head3)
result3 = odd_even_list(head3)
print_list("Test Case 3 (After):", result3)  # Output: 1 -> 3 -> 2 -> 4

# Test Case 4: Single node
head4 = ListNode(1)
print_list("Test Case 4 (Before):", head4)
result4 = odd_even_list(head4)
print_list("Test Case 4 (After):", result4)  # Output: 1

# Test Case 5: Empty list
head5 = None
print_list("Test Case 5 (Before):", head5)
result5 = odd_even_list(head5)
print_list("Test Case 5 (After):", result5)  # Output: None

File: 07_linked_list/07_07_odd_even_linked_list/odd_even_linked_list.cpp
#include <iostream>

class ListNode {
public:
    int val;
    ListNode* next;

    ListNode(int value = 0, ListNode* nextNode = nullptr) : val(value), next(nextNode) {}
};

ListNode* oddEvenList(ListNode* head) {
    if (!head || !head->next) {
        return head;
    }

    ListNode* odd = head;
    ListNode* even = head->next;
    ListNode* evenHead = head->next;
    while (even && even->next) {
        odd->next = odd->next->next;
        even->next = even->next->next;
        odd = odd->next;
        even = even->next;
    }

    odd->next = evenHead;
    return head;
}

// Helper function to print the linked list
void printList(const std::string& label, ListNode* head) {
    std::cout << label << " ";
    ListNode* current = head;
    while (current) {
        std::cout << current->val << " ";
        current = current->next;
    }
    std::cout << std::endl;
}

int main() {
    // Test Case 1: Odd and even number of nodes
    ListNode* head1 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5, new ListNode(6))))));
    printList("Test Case 1 (Before):", head1);
    ListNode* result1 = oddEvenList(head1);
    printList("Test Case 1 (After):", result1);  // Output: 1 -> 3 -> 5 -> 2 -> 4 -> 6

    // Test Case 2: Odd number of nodes
    ListNode* head2 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
    printList("Test Case 2 (Before):", head2);
    ListNode* result2 = oddEvenList(head2);
    printList("Test Case 2 (After):", result2);  // Output: 1 -> 3 -> 5 -> 2 -> 4

    // Test Case 3: Even number of nodes
    ListNode* head3 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4))));
    printList("Test Case 3 (Before):", head3);
    ListNode* result3 = oddEvenList(head3);
    printList("Test Case 3 (After):", result3);  // Output: 1 -> 3 -> 2 -> 4

    // Test Case 4: Single node
    ListNode* head4 = new ListNode(1);
    printList("Test Case 4 (Before):", head4);
    ListNode* result4 = oddEvenList(head4);
    printList("Test Case 4 (After):", result4);  // Output: 1

    // Test Case 5: Empty list
    ListNode* head5 = nullptr;
    printList("Test Case 5 (Before):", head5);
    ListNode* result5 = oddEvenList(head5);
    printList("Test Case 5 (After):", result5);  // Output: nullptr

    // Remember to free allocated memory
    delete head1;
    delete head2;
    delete head3;
    delete head4;

    return 0;
}
